File: cpp/examples/actionlib_msgs_pubsub.cpp
// actionlib_msgs_pubsub -- publishes and subscribes every
// actionlib_msgs type ReLink ships (relink/include/relink/
// standard_msgs.hpp), one topic per type: GoalID, GoalStatus,
// GoalStatusArray.
//
// Build:
//   g++ -std=c++17 -I relink/include -pthread examples/cpp/actionlib_msgs_pubsub.cpp -o actionlib_msgs_pubsub
// Run (in two terminals, or on two machines):
//   ./actionlib_msgs_pubsub

#include "relink/relink.hpp"
#include <cstdio>
#include <chrono>
#include <thread>

using namespace actionlib_msgs;

int main() {
    RelinkNode node;
    node.use_multicast_discovery();   // zero setup -- see Step 3

    node.subscribe<GoalID>("/relink/action/goal_id", [](const GoalID& m) {
        std::printf("actionlib_msgs/GoalID id=%s\n", m.id_str().c_str());
    });
    node.subscribe<GoalStatus>("/relink/action/goal_status", [](const GoalStatus& m) {
        std::printf("actionlib_msgs/GoalStatus status=%u text=%s\n", m.status, m.text_str().c_str());
    });
    node.subscribe<GoalStatusArray>("/relink/action/goal_status_array", [](const GoalStatusArray& m) {
        std::printf("actionlib_msgs/GoalStatusArray status_list_count=%u\n", m.status_list_count);
    });

    node.advertise<GoalID>("/relink/action/goal_id");
    node.advertise<GoalStatus>("/relink/action/goal_status");
    node.advertise<GoalStatusArray>("/relink/action/goal_status_array");

    int i = 0;
    while (true) {
        node.spin_once();   // services discovery -- call this every loop

        {
            GoalID gid{};
            gid.stamp = std_msgs::Time::now();
            char buf[32];
            std::snprintf(buf, sizeof(buf), "goal_%d", i);
            gid.set_id(buf);
            node.publish<GoalID>("/relink/action/goal_id", gid);
        }
        {
            GoalStatus gs{};
            gs.status = GoalStatus::kActive;
            gs.set_text("moving to goal");
            node.publish<GoalStatus>("/relink/action/goal_status", gs);
        }
        {
            GoalStatusArray gsa{};
            gsa.header.set_frame_id("action_server");
            gsa.status_list_count = 1;
            gsa.status_list[0].status = GoalStatus::kActive;
            node.publish<GoalStatusArray>("/relink/action/goal_status_array", gsa);
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(500));
        ++i;
    }
}

File: python/relink_py/examples/actionlib_msgs_pubsub.py
#!/usr/bin/env python3
"""actionlib_msgs_pubsub -- publishes and subscribes every
actionlib_msgs type ReLink ships (relink/standard_msgs.py), one topic
per type: GoalID, GoalStatus, GoalStatusArray. Wire-compatible with
cpp/examples/actionlib_msgs_pubsub.cpp.

Run (in two terminals, or on two machines):
    python3 actionlib_msgs_pubsub.py
"""
import sys
import os
import time

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))
from relink import RelinkNode
from relink import std_msgs, actionlib_msgs


def main():
    node = RelinkNode()
    node.use_multicast_discovery()   # zero setup -- see Step 3

    node.subscribe("/relink/action/goal_id", actionlib_msgs.GoalID, lambda m: print(f"actionlib_msgs/GoalID id={m.id_str()}"))
    node.subscribe("/relink/action/goal_status", actionlib_msgs.GoalStatus, lambda m: print(f"actionlib_msgs/GoalStatus status={m.status} text={m.text_str()}"))
    node.subscribe("/relink/action/goal_status_array", actionlib_msgs.GoalStatusArray, lambda m: print(f"actionlib_msgs/GoalStatusArray status_list_count={m.status_list_count}"))

    node.advertise("/relink/action/goal_id", actionlib_msgs.GoalID)
    node.advertise("/relink/action/goal_status", actionlib_msgs.GoalStatus)
    node.advertise("/relink/action/goal_status_array", actionlib_msgs.GoalStatusArray)

    i = 0
    while True:
        node.spin_once()   # services discovery -- call this every loop

        gid = actionlib_msgs.GoalID()
        gid.stamp = std_msgs.Time.now()
        gid.set_id(f"goal_{i}")
        node.publish("/relink/action/goal_id", gid)

        gs = actionlib_msgs.GoalStatus()
        gs.status = actionlib_msgs.GoalStatus.ACTIVE
        gs.set_text("moving to goal")
        node.publish("/relink/action/goal_status", gs)

        gsa = actionlib_msgs.GoalStatusArray()
        gsa.header.set_frame_id("action_server")
        gsa.status_list_count = 1
        gsa.status_list[0].status = actionlib_msgs.GoalStatus.ACTIVE
        node.publish("/relink/action/goal_status_array", gsa)

        time.sleep(0.5)
        i += 1


if __name__ == "__main__":
    main()
